presto-main/src/main/java/com/facebook/presto/type/FloatOperators.java

```diff
@@ -211,4 +214,44 @@ public static boolean castToBoolean(@SqlType(StandardTypes.FLOAT) long value)
     {
         return intBitsToFloat((int) value) != 0.0f;
     }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.BIGINT)
+    public static long saturatedFloorCastToBigint(@SqlType(StandardTypes.FLOAT) long value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p63f, -0x1p63f, Long.MAX_VALUE, Long.MIN_VALUE);
+    }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.INTEGER)
+    public static long saturatedFloorCastToInteger(@SqlType(StandardTypes.FLOAT) long value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p31f, -0x1p31f, Integer.MAX_VALUE, Integer.MIN_VALUE);
+    }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.SMALLINT)
+    public static long saturatedFloorCastToSmallint(@SqlType(StandardTypes.FLOAT) long value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p15f, -0x1p15f, Short.MAX_VALUE, Short.MIN_VALUE);
+    }
+
+    @ScalarOperator(SATURATED_FLOOR_CAST)
+    @SqlType(StandardTypes.TINYINT)
+    public static long saturatedFloorCastToTinyint(@SqlType(StandardTypes.FLOAT) long value)
+    {
+        return saturatedFloorCastToLong(value, 0x1p7f, -0x1p7f, Byte.MAX_VALUE, Byte.MIN_VALUE);
+    }
+
+    private static long saturatedFloorCastToLong(long valueBits, float maxValuePlusOneAsDouble, float minValueAsDouble, long maxValue, long minValue)
+    {
+        float value = intBitsToFloat((int) valueBits);
```

Comment:
same anwer ;-) But in addition not only the Long.MAX_VALUE can't be represented as Float. Integer.MAX_VALUE is too big to be represented precisely as Float as well.